src/architecture/unixfs.md

```diff
@@ -0,0 +1,611 @@
+this loses some of its meaning when mapped into most file systems today, as most file
+systems consider directories to be unordered key-value objects.
+
+A decoder SHOULD, if it can, preserve the order of the original files in the same way
+it consumed those names. However, when some implementations decode, modify and then
+re-encode, the original link order loses it's original meaning, given that there
+is no way to indicate which sorting was used originally.
+
+##### Path Resolution
+
+Pop the left-most component of the path, and try to match it to the `Name` of
+a child under `PBNode.Links`. If you find a match, you can then remember the CID.
+You MUST continue the search. If you find another match, you MUST error since
+duplicate names are not allowed. <!--TODO: check Kubo does this-->
+
+Assuming no errors were raised, you can continue to the path resolution on the 
+remaining components and on the CID you popped.
+
+
+#### `Symlink` type
+
+A :dfn[Symlink] represents a POSIX [symbolic link](https://pubs.opengroup.org/onlinepubs/9699919799/functions/symlink.html).
+A symlink MUST NOT have children. <!--TODO: check that this is true-->
+
+The `PBNode.Data.Data` field is a POSIX path that MAY be appended in front of the
+currently remaining path component stack.
+
+##### Path Resolution
+
+There is no current consensus on how pathing over symlinks should behave. Some
+implementations return symlink objects and fail if a consumer tries to follow them
+through.
+
+Following the POSIX specification over the current UnixFS path context is probably fine.
+
+#### `HAMTDirectory`
+
+A :dfn[HAMT Directory] is a [Hashed-Array-Mapped-Trie](https://en.wikipedia.org/wiki/Hash_array_mapped_trie)
+data structure representing a :ref[Directory]. It is generally used to represent
+directories that cannot fit inside a single block. These are also known as "sharded
+directories:, since they allow you to split large directories into multiple blocks, known as "shards".
+
+- `decode(PBNode.Data).hashType` indicates the [multihash] function to use to digest
+  the path components used for sharding. It MUST be `murmur3-x64-64` (`0x22`).
+- `decode(PBNode.Data).Data.Data` is a bit field, which indicates whether or not
+  links are part of this HAMT, or its leaves. The usage of this field is unknown, given
+  that you can deduce the same information from the link names.
+- `decode(PBNode.Data).Data.fanout` MUST be a power of two. This encodes the number
+  of hash permutations that will be used on each resolution step. The log base 2
+  of the `fanout` indicate how wide the bitmask will be on the hash at for that step.
+  `fanout` MUST be between 8 and probably 65536. <!-- 65536 is a totally arbitrary choice I made, FIXME: get consensus on an upper bound. -->.
+
+The field `Name` of an element of `PBNode.Links` for a HAMT starts with an
+uppercase hex-encoded prefix, which is `log2(fanout)` bits wide.
+
+##### Path Resolution
```

Comment:
TODO: cross check with GO and JS (https://github.com/ipfs/helia/blob/31cdfa8b8990acc9c99a55dd2c078d0c415055ea/packages/unixfs/src/commands/utils/dir-sharded.ts#L213C1-L225C1)